File: 199-binary-tree-right-side-view/binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        
        view = []
        self.traverseRight(root, 0, view)
        return view
        
    def traverseRight(self, root, depth, view):
        if root:
            if depth == len(view):
                view.append(root.val)
            
            self.traverseRight(root.right, depth + 1, view)
            self.traverseRight(root.left, depth + 1, view)

File: 199-binary-tree-right-side-view/binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public void dfs(TreeNode node, List<Integer> result, int height) {

        if (node == null){
            return;
        }

        if (result.size() == height) {
            result.add(node.val);
        }

        dfs(node.right, result, height + 1);
        dfs(node.left, result, height + 1);
    }

    public List<Integer> rightSideView(TreeNode root) {
        
        List<Integer> result = new ArrayList<>();
        dfs(root, result, 0);

        return result;
    }
}